states_rebuilder/lib/repository/todo_repository.dart

```diff
@@ -0,0 +1,46 @@
+import 'package:todos_repository_core/src/todo_entity.dart';
+import 'package:todos_repository_local_storage/todos_repository_local_storage.dart';
+
+import '../domain/entities/todo.dart';
+import '../service/exceptions/persistance_exception.dart';
+import '../service/interfaces/i_todo_repository.dart';
+
+class TodosRepository implements ITodosRepository {
```

Comment:
Could this be called something like `StatesRebuilderTodosRepository` and only accept a `TodosRepository` from the repository core instead of accepting a `KeyValueStore` and creating the `LocalStorageRepository` specifically? Right now this Repo is tied to those implementations, whereas I think it would be better to be tied to the abstract interface (TodosRepository).

That will make it easier to swap the Repository implementation in the Future to add support for other things like Desktop, which might not have a KeyValueStorage option and might need to rely on SQLite or File storage

Reply:
What is important is that the app (domain, service and UI) are decoupled from the concrete implementation of `TodosRepository`.
The idea in my mind is to make the domain and the service layers portable and independent even from the flutter_architecture_samples core.

Absolutely the concrete implementation of `TodosRepository` has to change if requirement changes. For example this is first implementation using FileStorage and a fake WebClient and before using the `keyValueStore`.

````dart
class TodoRepository implements ITodoRepository {
  final FileStorage fileStorage;
  final WebClient webClient;

  TodoRepository({this.fileStorage, this.webClient})
      : _todosRepositoryFlutter = TodosRepositoryFlutter(
          fileStorage: fileStorage,
          webClient: webClient,
        );

  final TodosRepositoryFlutter _todosRepositoryFlutter;
  @override
  Future<List<Todo>> loadTodos() async {
    try {
      final todoEntities = await _todosRepositoryFlutter.loadTodos();
      var todos = <Todo>[];
      for (var todoEntity in todoEntities) {
        todos.add(
          Todo.fromJson(todoEntity.toJson()),
        );
      }
      return todos;
    } catch (e) {
      throw PersistanceException('There is a problem in loading todos : $e');
    }
  }

  @override
  Future saveTodos(List<Todo> todos) {
    try {
      var todosEntities = <TodoEntity>[];
      for (var todo in todos) {
        todosEntities.add(TodoEntity.fromJson(todo.toJson()));
      }

      return _todosRepositoryFlutter.saveTodos(todosEntities);
    } catch (e) {
      throw PersistanceException('There is a problem in saving todos : $e');
    }
  }
}
```

Reply:
> What is important is that the app (domain, service and UI) are decoupled from the concrete implementation of TodosRepository.

Definitely. I guess what I'm asking: Could this class take in the `core.TodosRepository` as part of the constructor instead of a `KeyValueStore`? I understand the decoupling with the Arch Samples stuff, but in this case you're coupled to the concrete implementation of `core.TodosRepository`: `LocalStorageRepository`. It's also coupled to the `KeyValueStore` library. If this class just took in a `core.TodosRepsitory` that would allow me to just inject a different implementation in the future when I add desktop support, for example, while still maintaining the decoupling since it'd be relying on an abstract class instead of concrete implementation? :)

For example, if this class accepted a `core.TodosRepository` instead of the concrete `LocalStorageRepository` implementation, I could just inject one of the other repository implementations available. Right now we have `FileStorage` and `KeyValueStorage` options, and I'm working on adding `moor` and `hive` repositories, for example. Would be cool to just inject those into the app instead of having to refactor it later to accept those :)

Reply:
Yes, you are right. 
You mean some think like this:
```dart
class StatesRebuilderTodosRepository implements ITodosRepository {

  TodosRepository _todosRepository;

  StatesRebuilderTodosRepository({TodosRepository todosRepository})
      : _todosRepository =todosRepository;

  @override
  Future<List<Todo>> loadTodos() async {
    try {
      final todoEntities = await _todosRepository.loadTodos();
      var todos = <Todo>[];
      for (var todoEntity in todoEntities) {
        todos.add(
          Todo.fromJson(todoEntity.toJson()),
        );
      }
      return todos;
    } catch (e) {
      throw PersistanceException('There is a problem in loading todos : $e');
    }
  }

  @override
  Future saveTodos(List<Todo> todos) {
    try {
      var todosEntities = <TodoEntity>[];
      for (var todo in todos) {
        todosEntities.add(TodoEntity.fromJson(todo.toJson()));
      }

      return _todosRepository.saveTodos(todosEntities);
    } catch (e) {
      throw PersistanceException('There is a problem in saving todos : $e');
    }
  }
}
```

and in the main.dart file :

```dart
void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  runApp(
    StatesRebuilderApp(
      torepository: StatesRebuilderTodosRepository(
          todosRepository: KeyValueStorage(
        'states_rebuilder',
        FlutterKeyValueStore(await SharedPreferences.getInstance()),
      )),
    ),
  );
}
```

Reply:
Yep, exactly! That, or if you want to keep yours called `TodosRepository`, you can always name the import `core`, for example, and use `core.TodosRepository` :) Not sure if that's too confusing. 

This sample is also making me realize I should reorganize the repository code a bit more and extract some pieces out into a proper `domain` package -- so thanks very much for the discussion and sample!

Reply:
One more question: The current repositories are maybe better called "DataSources" in Clean Architecture terms -- do you think I should rename them to use that terminology instead?

Reply:
Yes, "DataSources" is a better name. I'm not very comfortable with the name repository, because it is the name of a design pattern that can be confusing. I think I will also use  'data_sources' to name the repository folder.